Confidential HGX H100/H200 guest: NVRC nv_fabricmanager aborts — topology file unreachable in composable base+gpu-extension layout (FM starting NvLink Inband panic)

## Summary

Confidential HGX H100/H200 guests (8x GPU + 4x NVSwitch, PPCIE) built with the **composable** kata-deploy image layout (driver-agnostic `nvidia` base + cold-plugged `nvidia-gpu-extension`) fail to bring up the NVSwitch fabric: the in-guest Fabric Manager aborts at init because its topology file is unreachable, and NVRC then panics with `timeout waiting for: FM starting NvLink Inband`. This blocks **all** GPU workloads on a confidential HGX NVSwitch guest (even single-GPU), not just multi-GPU NVLink.

The root cause is in **NVRC** (`NVIDIA/nvrc`), tracked upstream at **https://github.com/NVIDIA/nvrc/issues/203**. Filing here too because kata ships the composable images + pins NVRC, so kata users on HGX NVSwitch confidential guests hit this out of the box.

## Environment

- Kata Containers: main (Kata 4.0, `kata-deploy` OCI chart v4.0.0)
- NVRC: `v0.1.5` (pinned in `versions.yaml`)
- NVIDIA driver / fabricmanager: `595.58.03`
- Guest image layout (composable): boot image = `kata-ubuntu-noble-nvidia.image` (driver-agnostic base: NVRC init + kata-agent + base libs, **no GPU userspace**); cold-plugged `gpu` extension = `kata-ubuntu-noble-nvidia-gpu-extension-595.58.03.image` mounted at `/run/kata-extensions/gpu`
- Hardware: HGX Hopper (8x H200 + 4x NVSwitch `10de:22a3`), Intel TDX, PPCIE CC mode
- Runtime: `kata-qemu-nvidia-gpu-tdx-runtime-rs`

## Root cause (NVRC)

`NVIDIA/nvrc` `src/daemon.rs::nv_fabricmanager` copies **only** `fabricmanager.cfg` from the `gpu` extension to `/run/fabricmanager.cfg` and rewrites only `FABRIC_MODE` + `PARTITION_RAIL_POLICY`. The stock `fabricmanager.cfg` ships `TOPOLOGY_FILE_PATH=/usr/share/nvidia/nvswitch`. In the composable layout the booted **base** has **no** `/usr/share/nvidia/nvswitch/` (empty `/usr/share/nvidia/`); the topology files + cfg + `nv-fabricmanager` binary live **only** in the `gpu` extension at `/run/kata-extensions/gpu/usr/share/nvidia/nvswitch/`. NVRC never rewrites `TOPOLOGY_FILE_PATH` to that reachable extension path, so the FM aborts:

```
nv-fabricmanager: detected system topology is based on DGX/HGX H100
nv-fabricmanager: [ERROR] failed to open fabric topology file /usr/share/nvidia/nvswitch/dgxh100_hgxh100_topology
nv-fabricmanager: failed to open/parse fabric topology file information, aborting fabric manager
[  152.131265] NVRC panic: panicked at src/kmsg.rs:95:13:
[  152.131265] timeout waiting for: FM starting NvLink Inband
```

## What was needed to even reach this point (context)

Reaching the FM stage required two prior fixes that may be useful context for anyone debugging NVSwitch confidential guests:

1. **gpu-operator `vfio-manager` does not bind NVSwitches by default.** `k8s-driver-manager` PR #150 added `BIND_NVSWITCHES` (opt-in, default `false`). Without it, `vfio-manage bind --all` silently skips bridge-class NVSwitches (`0x0680`), so Kata pods requesting `nvidia.com/nvswitch` never schedule / NVRC panics earlier. Set `vfioManager.env.BIND_NVSWITCHES: "true"` in the gpu-operator values, and restart `nvidia-kata-sandbox-device-plugin` so it advertises `nvidia.com/nvswitch`.
2. **Kata `runtime-rs` filtered NVSwitches out of IOMMU groups + had no CDI kind for them.** PR #13431 (`core.rs` `0x0680` exception in `filter_bridge_device`; `cdi_devices/mod.rs` `CDI_DEVICE_KIND_TABLE` NVSwitch entry) fixes the runtime-rs path so NVSwitches pass through to the guest.

With both fixed, the sandbox creates and the NVSwitches appear in the guest (`nvidia-nvswitch0..3: open (major=240)`), and the FM is spawned — at which point the NVRC topology-file bug above is hit.

## Ask

- Track the upstream NVRC fix: **https://github.com/NVIDIA/nvrc/issues/203** (suggested fix: `configure_fabricmanager` should also rewrite `TOPOLOGY_FILE_PATH` to `gpu_extension::path("/usr/share/nvidia/nvswitch")` when the extension is present).
- Consider a kata-deploy-side mitigation: ship the `gpu` extension's `fabricmanager.cfg` with `TOPOLOGY_FILE_PATH=/run/kata-extensions/gpu/usr/share/nvidia/nvswitch` (the extension's own stable mount path) so the FM finds the topology files without an NVRC fix. This would require a rebuild of the `nvidia-gpu-extension` image + dm-verity recompute + `versions.yaml`/config `verity_params` `root_hash` update.

## Related

- Upstream NVRC bug: https://github.com/NVIDIA/nvrc/issues/203
- kata-containers#11912 — NVSwitch VFIO passthrough / NVLink unavailable
- kata-containers#13431 — runtime-rs NVSwitch IOMMU/CDI fix
- kata-containers#11730 — DGX H100 CUDA `system not yet initialized` (FM must run on HGX/DGX even for 1 GPU)

--- a/tools/osbuilder/rootfs-builder/nvidia/nvidia_rootfs.sh
+++ b/tools/osbuilder/rootfs-builder/nvidia/nvidia_rootfs.sh
@@ -592,6 +592,14 @@ partition_gpu_extension() {
 	# GPU configs (fabricmanager.cfg, nvlsm.conf, ...).
 	[[ -d usr/share/nvidia ]] && cp -a usr/share/nvidia/. "${extension}/usr/share/nvidia/"
 
+	# The topology files are only available under the GPU extension mount.
+	# Point Fabric Manager there so it can find them.
+	local fm_cfg="${extension}/usr/share/nvidia/nvswitch/fabricmanager.cfg"
+	if [[ -f "${fm_cfg}" ]]; then
+		sed -i 's|^TOPOLOGY_FILE_PATH=.*|TOPOLOGY_FILE_PATH=/run/kata-extensions/gpu/usr/share/nvidia/nvswitch|' \
+			"${fm_cfg}"
+	fi
+
 	# GPU firmware (GSP, ...); NVRC binds this onto /lib/firmware/nvidia.
 	[[ -d lib/firmware/nvidia ]] && cp -a lib/firmware/nvidia "${extension}/lib/firmware/"
 

